Evolution/0016-mutli-producer-single-consumer-channel.md

```diff
@@ -0,0 +1,847 @@
+strategy and both asynchronous and synchronous producers.
+
+### Multi/single consumer support
+
+The `AsyncSequence` protocol itself makes no assumptions about whether the
+implementation supports multiple consumers or not. This allows the creation of
+unicast and multicast asynchronous sequences. The difference between a unicast
+and multicast asynchronous sequence is if they allow multiple iterators to be
+created. `AsyncStream` does support the creation of multiple iterators and it
+does handle multiple consumers correctly. On the other hand,
+`AsyncThrowingStream` also supports multiple iterators but does `fatalError`
+when more than one iterator has to suspend. The original proposal states:
+
+> As with any sequence, iterating over an AsyncStream multiple times, or
+creating multiple iterators and iterating over them separately, may produce an
+unexpected series of values.
+
+While that statement leaves room for any behavior we learned that a clear distinction
+of behavior for root asynchronous sequences is beneficial; especially, when it comes to
+how transformation algorithms are applied on top.
+
+### Downstream consumer termination
+
+Downstream consumer termination allows the producer to notify the consumer that
+no more values are going to be produced. `Async[Throwing]Stream` does support
+this by calling the `finish()` or `finish(throwing:)` methods of the
+`Async[Throwing]Stream.Continuation`. However, `Async[Throwing]Stream` does not
+handle the case that the `Continuation` may be `deinit`ed before one of the
+finish methods is called. This currently leads to async streams that never
+terminate.
+
+### Upstream producer termination
+
+Upstream producer termination is the inverse of downstream consumer termination
+where the producer is notified once the consumption has terminated. Currently,
+`Async[Throwing]Stream` does expose the `onTermination` property on the
+`Continuation`. The `onTermination` closure is invoked once the consumer has
+terminated. The consumer can terminate in four separate cases:
+
+1. The asynchronous sequence was `deinit`ed and no iterator was created
+2. The iterator was `deinit`ed and the asynchronous sequence is unicast
+3. The consuming task is canceled
+4. The asynchronous sequence returned `nil` or threw
+
+`Async[Throwing]Stream` currently invokes `onTermination` in all cases; however,
+since `Async[Throwing]Stream` supports multiple consumers (as discussed in the
+`Multi/single consumer support` section), a single consumer task being canceled
+leads to the termination of all consumers. This is not expected from multicast
+asynchronous sequences in general.
+
+## Proposed solution
+
+The above motivation lays out the expected behaviors for any consumer/producer
+system and compares them to the behaviors of `Async[Throwing]Stream` and
+`Async[Throwing]Channel`.
+
+This section proposes a new type called `MultiProducerSingleConsumerAsyncChannel`
```

Comment:
Id really love to come up with a smaller name. We have been using water metaphors, so perhaps something in that family of names? From my understanding of the pitch it is like you have many small tributaries contributing to a larger stream of values. Id like to come up with some alternatives than this.

Reply:
4 words is better than 6 I guess...